clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java

```diff
@@ -632,26 +633,71 @@ public void testSendToInvalidTopic() throws Exception {
-                                                            Collections.<String>emptySet(),
-                                                            invalidTopic,
-                                                            cluster.internalTopics(),
-                                                            cluster.controller());
+                cluster.nodes(),
+                new ArrayList<PartitionInfo>(0),
+                Collections.<String>emptySet(),
+                invalidTopic,
+                cluster.internalTopics(),
+                cluster.controller());
         client.prepareMetadataUpdate(metaDataUpdateResponseCluster, Collections.<String>emptySet());
 
         Future<RecordMetadata> future = producer.send(record);
 
         assertEquals("Cluster has incorrect invalid topic list.", metaDataUpdateResponseCluster.invalidTopics(), metadata.fetch().invalidTopics());
         TestUtils.assertFutureError(future, InvalidTopicException.class);
     }
+
+    @Test
+    public void testCloseWhenWaitingForMetadataUpdate() throws ExecutionException, InterruptedException {
+        Properties props = new Properties();
+        props.put(ProducerConfig.MAX_BLOCK_MS_CONFIG, Long.MAX_VALUE);
+        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9000");
+
+        // Simulate a case where metadata for a particular topic is not available. This will cause KafkaProducer#send to
+        // block in Metadata#awaitUpdate for the configured max.block.ms. When close() is invoked, KafkaProducer#send should
+        // return with a KafkaException.
+        Time time = new MockTime();
+        Cluster cluster = TestUtils.singletonCluster();
+        Node node = cluster.nodes().get(0);
+        Metadata metadata = new Metadata(0, Long.MAX_VALUE, false);
+        metadata.update(cluster, Collections.<String>emptySet(), time.milliseconds());
+        MockClient client = new MockClient(time, metadata);
+        client.setNode(node);
+
+        Producer<String, String> producer = new KafkaProducer<>(
+                new ProducerConfig(ProducerConfig.addSerializerToConfig(props, new StringSerializer(), new StringSerializer())),
+                new StringSerializer(), new StringSerializer(), metadata, client);
+
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        final AtomicReference<Exception> sendException = new AtomicReference<>();
+
+        try {
+            executor.submit(() -> {
+                try {
+                    // Metadata for topic "test" will not be available which will cause us to block indefinitely until
+                    // KafkaProducer#close is invoked.
+                    producer.send(new ProducerRecord<>("test", "key", "value"));
+                    fail();
+                } catch (Exception e) {
+                    sendException.set(e);
+                }
+            });
+
+            Thread.sleep(100);
```

Comment:
This feels a little brittle. If there is a delay before executing the task, then send() may raise the wrong exception. I think we could make it more reliable by waiting until the topic "test" has been added to `Metadata`.

Reply:
Good point. There is still some degree of uncertainty (even if much smaller than before) so I retained the sleep.

Reply:
Can you elaborate? I don't see any point in the code where we would return between adding the topic and awaiting the update.